### ISSUE TITLE
Caching large packages craches server

### ISSUE THREAD

**Issue by mamoit** [stevearc/pypicloud#210]
I'm trying to install `torch` (over 500MB) through a `pypicloud` instance (as cache layer), but it always crashes with what looks like either a timeout or a memory allocation issue.
Every other (smaller) package works like a treat.
Is the package cached in memory while it's being downloaded from upstream for caching purposes?

```
spawned uWSGI http 1 (pid: 26)
[pid: 23|app: 0|req: 1/1] 127.0.0.1 () {34 vars in 803 bytes} [Wed Apr 24 17:30:46 2019] GET /simple/torch/ => generated 8700 bytes in 212 msecs (HTTP/1.1 200) 2 headers in 81 bytes (1 switches on core 0)
INFO 2019-04-24 17:30:47,631 [pypicloud.views.api] Caching torch-1.0.1.post2-cp36-cp36m-manylinux1_x86_64.whl from https://pypi.python.org/simple
*** /etc/my_init.d/pypicloud-uwsgi.sh failed with status 0

*** Killing all processes...
SIGINT/SIGQUIT received...killing workers...
gateway "uWSGI http 1" has been buried (pid: 26)
Traceback (most recent call last):
  File "/sbin/my_init", line 417, in <module>
    main(args)
  File "/sbin/my_init", line 340, in main
    run_startup_files()
  File "/sbin/my_init", line 260, in run_startup_files
    run_command_killable_and_import_envvars(filename)
  File "/sbin/my_init", line 221, in run_command_killable_and_import_envvars
    run_command_killable(*argv)
  File "/sbin/my_init", line 217, in run_command_killable
    sys.exit(1)
SystemExit: 1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/sbin/my_init", line 423, in <module>
    kill_all_processes(KILL_ALL_PROCESSES_TIMEOUT)
  File "/sbin/my_init", line 238, in kill_all_processes
    os.waitpid(-1, 0)
  File "/sbin/my_init", line 413, in <lambda>
    signal.signal(signal.SIGTERM, lambda signum, frame: ignore_signals_and_raise_keyboard_interrupt('SIGTERM'))
  File "/sbin/my_init", line 61, in ignore_signals_and_raise_keyboard_interrupt
    raise KeyboardInterrupt(signame)
KeyboardInterrupt: SIGTERM
```

my config looks something like this:
```
[app:main]
use = egg:pypicloud

pyramid.reload_templates = false
pyramid.debug_authorization = false
pyramid.debug_notfound = false
pyramid.debug_routematch = false
pyramid.default_locale_name = en

pypi.default_write = everyone
pypi.default_read = everyone
pypi.cache_update = everyone

pypi.storage = gcs
storage.bucket = somebucket
pypi.fallback = cache
pypi.auth = sql

db.url = sqlite:////var/lib/pypicloud/db.sqlite
auth.db.url = sqlite:////var/lib/pypicloud/db.sqlite

[uwsgi]
paste = config:%p
paste-logger = %p
master = true
uid = pypicloud
gid = pypicloud
processes = 5
reload-mercy = 15
harakiri = 30
worker-reload-mercy = 15
max-requests = 1000
http = 0.0.0.0:8080
```

I've tried with `enable-threads = true` too, but it just kills the threads over and over again too.

**Comment by stevearc** [comment]
Yes, it's reading the entire file into memory in order to both upload it to the storage backend and send it down the the client.

https://github.com/stevearc/pypicloud/blob/feae61f02d3abd734d771f4a74159b460e98063a/pypicloud/views/api.py#L71-L78

I tried it locally in a docker container and it seemed to work fine. Could you try using the filesystem storage backend and see if you get the same crash? Another thing to check would be any resource limits set on the container.

**Comment by mamoit** [comment]
I haven't tried with local storage yet.
My crash is most likely due to the RAM constraints that I have imposed to the container.
I'm running it in k8s, with `1Gi` of memory as a limit.
I'll try raising it and see if it solves the problem.
Would storing it in a tmp file instead of in RAM be a desirable feature?

**Comment by mamoit** [comment]
When you cast the `data` to `BytesIO` doesn't it duplicate the object?
That explains why having 1Gi of RAM is not enough to store 500Mi of package.

**Comment by stevearc** [comment]
I don't believe that's it. `BytesIO` is just storing a reference to the object
https://github.com/python/cpython/blob/6d13e5b35ba7165db3f38fccdd6e1e5283f96a74/Modules/_io/bytesio.c#L921-L926

I did some testing and it seems like Pyramid isn't releasing the memory used by the response object. I tried monitoring the memory while going through the cache+serve flow, and the memory jumps to ~600M and stays there. Repeating thecall only adds to it. This isn't only true of caching; if you set `pypi.stream_files = true` it does the same thing every time it serves the torch package. The common code here is that we're taking a big ol' blob of `data` and setting `response.body = data`.
